### ISSUE TITLE
Faster RCNN on Own Dataset

### ISSUE THREAD

**Issue by xXBasti** [pytorch/vision#986]
Hi all,

I am trying to train a modification of the Faster RCNN with a changed Head on my own dataset.
I followed the tutorial on https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html.

As I want to use the Faster RCNN for detection only, I changed to part for the data loader:
So that __getitems__ returns:
        target["boxes"] = #coordinates of the bounding box between 0 and H or W
        target["labels"] = #label id from 0 to 43 (I have 43 classes)
        target["image_id"] = # as described torch.tensor([idx])
        target["area"] = #as described
        target["iscrowd"] = #as described

After changing the the classes and remove the mask, i tried to train the model with a different backbone like approach 2. 

Now I am getting after every epoch just zeros, doesn't matter if I am using the suggested SGD, RMSProp or Adam.

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000

I wonder if my data format is wrong or there is something I missing there. 
On the other hand I tried to debug, the code and check the tensors beeing processed. I noticed that the features as well as the proposals are close to zero.

The first two step look like this:
Epoch: [0]  [  0/640]  eta: 0:05:31  lr: 0.000128  loss: 4.5273 (4.5273)  loss_classifier: 3.8356 (3.8356)  loss_box_reg: 0.0000 (0.0000)  loss_objectness: 0.6894 (0.6894)  loss_rpn_box_reg: 0.0023 (0.0023)  time: 0.5178  data: 0.1876  max mem: 2013
Epoch: [0]  [ 10/640]  eta: 0:02:10  lr: 0.000910  loss: 0.7780 (1.0823)  loss_classifier: 0.5873 (0.7998)  loss_box_reg: 0.0242 (0.0431)  loss_objectness: 0.1112 (0.2068)  loss_rpn_box_reg: 0.0182 (0.0326)  time: 0.2074  data: 0.0207  max mem: 2889

It would be nice, if someone can help me and tell me my mistake using Faster RCNN for detection.

Kind Regards 
xXBasti

**Comment by fmassa** [comment]
Note that if you changed the backbone, it will take some time (and a lot of data) before it actually starts to learn something.

So that's probably the reason why it's not working at all for you.
Using transfer learning (with the same backbone) trains much faster

**Comment by xXBasti** [comment]
Thanks for your help. 

Yeah true, changing the backbone requires more training time and data. But I am still using a pretrained backbone. And sometimes it gives some reasonable region proposals.
But it is even impossible to overfit the model on the trainings data. The evaluation score on the trainings set is still 0 in all categories and the model does not detect a single object. 

Now a also tried to use transfer learning. This seems working, but when I replace the MLPHead and the box predictor, I the only different between using a pretrained backbone and a pretrained FasterRCNN is the RPN, isn't it?

So I am still wondering why it wasn't working before.
But for my experiments I can also use the pretrained fasterrcnn_resnet50_fpn.

**Comment by fmassa** [comment]
Using a pre-trained backbone (trained on classification) and one which has been trained for detection is not the same thing.

> This seems working, but when I replace the MLPHead and the box predictor, I the only different between using a pretrained backbone and a pretrained FasterRCNN is the RPN, isn't it?

If I understand your points correctly, then yes. But having a good RPN is already very helpful in the beginning of training, which might indicate why you start with better results in the beginning.

I believe you have fixed your issue now, so I'm closing it. Let me know if you have further questions

**Comment by xXBasti** [comment]
Ok, I thought they were also trained on classification.

Yes, everything clear, thanks for the help!

**Comment by mkport1** [comment]
Hello, 

I'm trying as well to implement Faster RCNN on my own dataset and I'm facing the same issue. Can you please provide your code for the model @xXBasti (the replacement for  MLPHead and the box predictor  part especially)

**Comment by fmassa** [comment]
@mkport1 check https://colab.research.google.com/github/pytorch/vision/blob/temp-tutorial/tutorials/torchvision_finetuning_instance_segmentation.ipynb , it has a section on how to build a Faster R-CNN model

**Comment by icesohelrana** [comment]
Hi @fmassa ,
I have a similar problem. I have used approximately 5000 images to train with a pre-trained backbone of fasterrcnn_resnet50_fpn. In the first epoch, it's showing some good result like this. 
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.293
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.639
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.223
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.011
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.253
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.453
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.033
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.259
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.391
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.100
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.364
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.531
 
But from next epoch, all accuracy goes to zero. How can I solve this?

**Comment by fmassa** [comment]
Without further information it is very hard to identify what might be the issue.
Maybe your training data has a problem somewhere?

**Comment by 17sarf** [comment]
@icesohelrana did you manage to find the issue that was causing the accuracy to go down to zero? I am currently having the same problem. 

@fmassa I am trying to fine-tune a model on the CityPersons dataset. I have checked my data by viewing the images and drawing the bounding boxes on the relevant objects, and everything seems to be fine. To start off, I want to simply train the model on a single class. 

My model is set up like this:

```
backbone = torchvision.models.resnet50(pretrained=True)
backbone = torch.nn.Sequential(*list(backbone.children()))[:-2]
backbone.out_channels = 2048

anchor_generator = AnchorGenerator(sizes=((32, 64, 128, 256, 512),),
                                   aspect_ratios=((0.5, 1.0, 2.0),))

roi_pooler = torchvision.ops.MultiScaleRoIAlign(featmap_names=['0'],
                                                output_size=7,
                                                sampling_ratio=2)

model = FasterRCNN(backbone,
                   num_classes=2,
                   rpn_anchor_generator=anchor_generator,
                   box_roi_pool=roi_pooler)
```

Any advice or suggestions would be greatly appreciated.

Thank you in advance.

**Comment by GKalliatakis** [comment]
I'm facing the same issue (getting zeros for every evaluation step). 

I'm working with grayscale images (and this is taken into consideration by doing the following: 

```python
model.backbone.body.conv1 = torch.nn.Conv2d(in_channels, 64, kernel_size=7, stride=2, padding=3, bias=False)

grcnn = GRCNNTransform(min_size=1024, max_size=1024, image_mean=[0.485], image_std=[0.229])
model.transform = grcnn

```

Any idea what's going wrong?

**Comment by 17sarf** [comment]
@GKalliatakis the issue I found for my code was that there was an error in the dataset/dataloader. I can't remember exactly what it was, but I think it was something to do with the way I was pre-processing the data. Sorry I couldn't be more help.

**Comment by GKalliatakis** [comment]
Thanks @fmassa for your reply. I've tried using RGB images instead of grayscale but performance is similar. I can see a very small improvement after many epochs (fasterrcnn_resnet50_fpn pre-trained on COCO) but nowhere close to reported mAP. I guess the number of training samples is not enough even for 10,250 trainable params (alongside the different nature of my images).

**Comment by datumbox** [comment]
> I guess the number of training samples is not enough even for 10,250 trainable params (alongside the different nature of my images).
@GKalliatakis Quite possibly that's the case. 

As a last-ditch effort try also using the original mean/std (along with the 3-channel trick). The model is trained with the specific values and since you don't fine-tune the weights on the bottom of the network you might get issues.
